prow/gerrit/client/client.go

```diff
@@ -366,72 +382,86 @@ func (h *gerritInstanceHandler) queryChangesForProject(project string, lastUpdat
-					if !ok {
-						logrus.WithError(err).Errorf("(should not happen?)cannot find current revision for change %v", change.ID)
-						continue
-					}
+			log := log.WithFields(logrus.Fields{
+				"change":     change.Number,
+				"updated":    change.Updated,
+				"status":     change.Status,
+				"lastUpdate": lastUpdate,
+			})
 
-					created := parseStamp(rev.Created)
-					h.injectPatchsetMessages(&change)
-					changeMessages := change.Messages
-					var newMessages bool
-
-					for _, message := range changeMessages {
-						if message.RevisionNumber == rev.Number {
-							messageTime := parseStamp(message.Date)
-							if messageTime.After(lastUpdate) {
-								logrus.Infof("Change %d: Found a new message %s at time %v after lastSync at %v", change.Number, message.Message, messageTime, lastUpdate)
-								newMessages = true
-								break
-							}
-						}
-					}
+			// stop when we find a change last updated before lastUpdate
+			if !updated.After(lastUpdate) {
+				log.Info("No more recently updated changes")
+				return pending, nil
+			}
+
+			// process recently updated change
+			switch change.Status {
+			case Merged:
+				submitted := parseStamp(*change.Submitted)
+				log := log.WithField("submitted", submitted)
+				if !submitted.After(lastUpdate) {
+					log.Info("Skipping previously merged change")
+					continue
+				}
+				log.Info("Found merged change")
+				pending = append(pending, change)
+			case New:
+				// we need to make sure the change update is from a fresh commit change
+				rev, ok := change.Revisions[change.CurrentRevision]
+				if !ok {
+					log.WithError(err).WithField("revision", change.CurrentRevision).Error("Revision not found")
+					continue
+				}
 
-					if !newMessages && !created.After(lastUpdate) {
-						// stale commit
-						logrus.Infof("Change %d, latest revision updated %s before lastUpdate %s, skipping this patchset", change.Number, created, lastUpdate)
-						continue
+				created := parseStamp(rev.Created)
+				log := log.WithField("created", created)
+				h.injectPatchsetMessages(&change)
+				changeMessages := change.Messages
+				var newMessages bool
+
+				for _, message := range changeMessages {
+					if message.RevisionNumber == rev.Number {
+						messageTime := parseStamp(message.Date)
+						if messageTime.After(lastUpdate) {
+							log.WithFields(logrus.Fields{
+								"message":     message.Message,
+								"messageDate": messageTime,
+							}).Info("New messages")
+							newMessages = true
+							break
+						}
 					}
+				}
 
-					pending = append(pending, change)
-				default:
-					// change has been abandoned, do nothing
+				if !newMessages && !created.After(lastUpdate) {
+					// stale commit
+					log.Info("Skipping existing change")
+					continue
 				}
-			} else {
-				logrus.Infof("Change %d, updated %s before lastUpdate %s, return", change.Number, change.Updated, lastUpdate)
-				return pending, nil
+				if !newMessages {
```

Comment:
The log message is opposite from what's meant by this if check, should it be `if newMessages`?

Reply:
If there are new messages then we got to L447 which means we already logged on line L443.

L443 covers the case where the change was updated by a comment.

L459 covers the case where someone pushed new code.

Reply:
Got it. Thanks for the explanation!